crypto/src/key/secp256k1/extended_keys.rs

```diff
@@ -0,0 +1,256 @@
+// Copyright (c) 2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::key::hdkd::chain_code::CHAINCODE_LENGTH;
+use crate::key::hdkd::{
+    chain_code::ChainCode,
+    child_number::ChildNumber,
+    derivable::{Derivable, DerivationError},
+};
+use crate::key::secp256k1::{Secp256k1PrivateKey, Secp256k1PublicKey};
+use crate::random::{CryptoRng, Rng};
+use hmac::{Hmac, Mac};
+use secp256k1;
+use secp256k1::SECP256K1;
+use serialization::{Decode, Encode};
+use sha2::Sha512;
+use zeroize::Zeroize;
+
+// Create alias for HMAC-SHA512
+type HmacSha512 = Hmac<Sha512>;
+
+#[derive(Debug, PartialEq, Eq, Clone, Encode, Decode)]
+pub struct Secp256k1ExtendedPrivateKey {
+    /// Chain code
+    pub chain_code: ChainCode,
+    /// Private key
+    pub private_key: Secp256k1PrivateKey,
+}
+
+impl Secp256k1ExtendedPrivateKey {
+    pub fn new_master(seed: &[u8]) -> Result<Secp256k1ExtendedPrivateKey, DerivationError> {
+        let mut mac =
+            HmacSha512::new_from_slice(b"Bitcoin seed").expect("HMAC can take key of any size");
```

Comment:
What's the purpose of the `"Bitcoin seed"` string here. Is that for compatibility so that we generate identical key tree to Bitcoin?

Reply:
Yes, it's: https://en.bitcoin.it/wiki/BIP_0032#Master_key_generation

But I'm wondering now whether we want to parametrize this for our own use. Is this a standard that everyone uses in every crypto? Or is it just in bitcoin and everyone adapts it to whatever they want?

Reply:
The second hardened component is usually a coin type in the usual layout. That way you can have a wallet that communicates with multiple networks and is able to generate keys for all from the same seed. Since the Mintlayer wallet is supposed to interface both Mintlayer and Bitcoin, I think we should just use the `"Bitcoin seed"` string and allocate a new number for the second derivation path component. Unless there are other complications I have not considered, that is.

https://learnmeabitcoin.com/technical/derivation-paths#wallet-structure

Reply:
The "Bitcoin seed" constant is defined in BIP39 and it is used for most of the coins. Some coins do change it but it is when they use different elliptic curve. I our case we don't need to change it.